### PR TITLE
I14 assigning competencies

### LINKED ISSUE #14: assign competencies to student based on the criteria by course design
Parent issue #7 
This issue deals with assigning competencies to students in the course. 
1. Only active student enrollments are gathered
2. based on the criteria assigned for a course in the google spreadsheet,[All Enrolled or > 70% grade], students will receive competencies
3.  final_score is consider for competency assignment

Tests are written testing most of the scenarios. 
It's is taking about 2 min to assign competencies for ~200 students in the course

Screen Shot 2021-11-18 at 3 56 14 PM

Logging format for assigning competencies
Success Case
```
INFO: For course 448447 with competencies criteria: `Passing - 70% or Higher`, 5/5 students will receive competencies
INFO: 5/5 Students assigned competencies in course 448447 in assignment 448447122
```
Few Failure case
```
INFO: For course 448447 with competencies criteria: `All Enrolled`, 10 students will receive competencies 
ERROR: 7/10 Students didn't got competancies assigned in course 448447 in assignment 448447122 to the following students [{136950: 'Student User6'}, {136951: 'Student User7'}, {136953: 'Student User9'}]
```

### PR DESCRIPTION
**Pull request by pushyamig**
Fixes #14 and this resolves parent issue #7 

Use below google file for testing. I have only configured 3 courses [448447, 454941, 484176] in the spreadsheet for making the testing short. 

 https://docs.google.com/spreadsheets/d/18mtjJ8RRP5MnSQYqzhJtwVBmdkFPy3m0u_i_Rkmuk0A/edit#gid=0

### CODE REVIEW

**Comment by zqian** [comment]
I got the following error message for both course #448447 and course #454941

job_1  | ipe_process_orchestrator.assign_competencies - INFO - For course 454941 with competencies criteria: 'All Enrolled', 48 students will receive competencies
job_1  | api_handler.api_calls - ERROR - Received irregular status code: 500
job_1  | api_handler.api_calls - ERROR - {"errors":[{"message":"An error occurred.","error_code":"internal_server_error"}],"error_report_id":xxxxx}
job_1  | api_handler.api_calls - INFO - Beginning next_attempt

Not sure what happened. 

Could the log message include more details?

File: ipe_process_orchestrator/orchestrator.py [hunk 1]
```diff
@@ -75,22 +68,43 @@ def getting_rubrics(self):
...
+        """
+        Check if the competencies values are given in the Google Sheet. If not, then exit the process.
+        Not a real life usecase, just checking for sanity.
+        """
+        try:
+            if(course[COL_COMPETENCIES_RR] and course[COL_COMPETENCIES_TTW] and course[COL_COMPETENCIES_IC] and
+               course[COL_COMPETENCIES_VE] and course[COL_COMPETENCIES_IH] and course[COL_DOSAGE] and course[COL_ASSIGNING_LO_CRITERIA]):
+                return True
+            else: 
+              logger.error(f'Competencies few or all values not provided in the Google Sheet so skipping competency process for course: {course[COL_COURSE_ID]}')
```
Comment (zqian) [inline ipe_process_orchestrator/orchestrator.py]: Maybe a rewording is needed in this error message "Competencies few or all values not provided in..."?

something like "Not all required competencies values are provided in..."

File: ipe_process_orchestrator/orchestrator.py [hunk 2]
```diff
@@ -75,22 +68,43 @@ def getting_rubrics(self):
...
+              return False
+        except Exception as e:
+            logger.error(f'Error in getting the ipe competencies values from Google Sheet so skipping competency process for course {course[COL_COURSE_ID]}: {e}')
+            return False
+
+    def start_competencies_assigning_process(self, course: pd.Series, rubric_data) -> None:
         """
         First step in the assiging competencies process is to create the asssignment if it does not exist.
         Second step is to assign competencies to the assignment.
+        Thrid step is to update the status in Google Sheets that the competencies are assigned.
```
Comment (zqian) [inline ipe_process_orchestrator/orchestrator.py]: typo for "Third"

File: ipe_process_orchestrator/orchestrator.py [hunk 3]
```diff
@@ -75,22 +68,43 @@ def getting_rubrics(self):
...
         """
         try:
-            assignment_id = self._create_delete_assignment(course)
+            if not self.check_competencies_values_given_gsheet(course):
```
Comment (zqian) [inline ipe_process_orchestrator/orchestrator.py]: this if statement can be outside of the try-except.

Comment (pushyamig) [reply ipe_process_orchestrator/orchestrator.py]: Done

File: ipe_process_orchestrator/orchestrator.py (hunk 2 above)
Comment (pushyamig) [reply ipe_process_orchestrator/orchestrator.py]: Done

File: ipe_process_orchestrator/orchestrator.py (hunk 1 above)
Comment (pushyamig) [reply ipe_process_orchestrator/orchestrator.py]: Done

**Comment by pushyamig** [comment]
> Not sure what happened.
Could the log message include more details?

We discussed this over zoom session. So I am thinking this is resolved.

**Comment by pushyamig** [comment]
@zqian I have resolved all the comments. Please review again.

**Review by zqian — APPROVED** [review]

**Pull request merged**